### DISCUSSION TITLE
Auto select closest matched entry for cmdline and LSP?

### DISCUSSION THREAD

**Discussion by rieje** [hrsh7th/nvim-cmp#1480] [Q&A] [answered]
(Disclaimer: new to Neovim, not sure if this is Neovim or nvim-cmp related.)

How to auto select closest matched entry for cmdline and LSP? To be clear, I type a partial match and just hit Enter to complete it. Also, the option to fill the text with that closest match if I did not explicitly select it would be nice. With my minimal config which I assume is the default since I haven't found the related setting, I need a Ctrl-n to select the closest match at the top of the list and then an Enter. To me, autoselecting the closest match by default seems to make most sense because user can just hit Enter to select it or quick completion.

**Comment by fitrh** [comment]
What do you mean by **closest match**? The first entry?

You can try `confirm` with `select = true`

```lua
["<CR>"] = cmp.mapping.confirm({ select = true })
```

Or set `completion.completeopt`

```lua
cmp.setup({
  -- ...
  completion = {
    completeopt = "menu,menuone,noinsert",
  }
  -- ...
})
```

**Reply by rieje** [reply]
Yes, I should be able to just hit Enter after typing something partial instead of Ctrl-n to select the item. For some reason it never selects anything on the list, including the first entry (i.e. closest match).

**Reply by fitrh** [reply]
I'm not sure if I understand what you want, but you can try enable the `select` opts for `confirm()`, something like this 

```lua 
["<CR>"] = cmp.mapping.confirm({ select = true })
```

**Reply by rieje** [reply]
That's what I have. To clarify:

[This is the behavior I want (from Emacs)](https://0x0.st/HiAG.png) where after typing the string if I were to hit enter it selects the closest match at the top that's *already highlighted".

In Neovim, [this is what I get](https://0x0.st/HiAD.png)--the closest match does not get highlighted. I expect NvimTreeOpen to be highlighted so that hitting enter uses it. It only gets selected if I do a Ctrl-n, otherwise nothing gets selected and hitting enter results in "E464:Ambiguous use of user-defined command". I.e. is no preselection for closest match which should be the top item on the list.

**Reply by fitrh** [reply]
Did you set `completion.completeopt`?

```lua
cmp.setup({
  -- ...
  completion = {
    completeopt = "menu,menuone,noinsert",
  }
  -- ...
})
```

And I don't think the `<CR>` mapping causes that kind of error

**Reply by rieje** [reply]
Yea that was it, just made an answer as you post this, I'll mark yours as the answer. Thanks.

**Reply by fitrh** [reply]
Glad you found it

BTW, I still convince that the `<CR>` mapping should not cause that error, and `select` opts will select the first entry even if preselection is disabled

Here's my config, I don't use preselect but use the `select` opts


https://user-images.githubusercontent.com/55179750/224216258-9b724661-5c73-4c87-a80a-07d7b8cf08e4.mp4

**Comment by rieje** [comment, marked as answer]
[Answer](https://github.com/hrsh7th/nvim-cmp/discussions/1480#discussioncomment-5262554) (I can't mark non-top level comment as answer).
